Fix floating label text cut off in select elements

theodorejb opened the pull request:
### Description

Tweaks line height and padding so that letters with descenders will not be cut off when using a floating label on a `<select>` element.

### Motivation & Context

In Bootstrap 5.3.3 floating labels cause the bottom of some letters like j, g, p, and y to be cut off. This occurs with normal `form-select` dropdowns in Firefox and Chrome/Edge, and the problem gets significantly worse when using `form-select-lg`, such that this class is unusable with floating labels in all browsers (including Safari).

- Demo: https://stackblitz.com/edit/bs-select-floating-label?file=index.html
- Screenshot:
![Edge screenshot of BS floating label bug](https://github.com/twbs/bootstrap/assets/3053271/b777ba29-81f3-4b9c-be7b-dd0b5bef7358)

### Type of changes

- [x] Bug fix (non-breaking change which fixes an issue)

### Checklist

- [x] I have read the [contributing guidelines](https://github.com/twbs/bootstrap/blob/main/.github/CONTRIBUTING.md)
- [x] My code follows the code style of the project _(using `npm run lint`)_
- [ ] My change introduces changes to the documentation

#### Live previews

- <https://deploy-preview-39483--twbs-bootstrap.netlify.app/docs/5.3/forms/floating-labels/#selects>

### Related issues

Fixes #33711, fixes #34147

theodorejb committed:
Fix floating label text cut off in select elements

Fixes #33711, fixes #34147

theodorejb commented:
@julien-deramond Is there any chance this fix can be included in v5.3.3?

mdo commented:
Unfortunately I think these are somewhat arbitrary values that make our components a little more cumbersome to tweak. I'd rather see us go the route of an improved, balanced vertical padding for both top and bottom vs separate values for both.

theodorejb commented:
@mdo Do you mean using the same top/bottom padding for both selects and inputs? I tried that first, and while the new values do work fine for input elements, they do not work for textareas (it causes the label to overlap the textarea contents).

If there's a way to fix the floating select label bug without new variables, that would be great! But I spent quite a bit of time playing around with different values/approaches, and this was the simplest one I found that worked for select elements in all cases without breaking anything else.

theodorejb commented:
I changed the approach to use consistent padding for input and select elements, and only override it for textareas. Also I removed the unnecessary floating label background for non-textareas, so this merge request can supersede #39480.

Note that the 1px increase to line height is necessary to fix the floating label text cutoff issue for select dropdowns. But I'm not sure if this line height increase is desirable for textareas (multi-line text uses slightly more vertical space). If needed another variable could be added to keep the line height at 1.25 for textareas.

theodorejb commented:
I rebased this on top of #39720 since that was approved. I hope this can also make it into a bugfix release.

louismaximepiton commented on scss/_variables.scss:
```diff
@@ -1069,11 +1069,13 @@ $form-file-button-hover-bg:       var(--#{$prefix}secondary-bg) !default;
 // scss-docs-start form-floating-variables
 $form-floating-height:                  add(3.5rem, $input-height-border) !default;
-$form-floating-line-height:             1.25 !default;
+$form-floating-line-height:             1.375 !default;
```
I think I would keep the previous value for the line-height so it won't break anything for people and keep the changes very tight to the initial problem.

louismaximepiton replied:
Ah yes my bad, I didn't test on Chrome but on FF and it was causing no issue.

louismaximepiton commented on scss/forms/_floating-labels.scss:
Just a remark, but I think that CSS variables are a bit overkill on this. Maybe we can keep them as progressive enhancement of the lib.

louismaximepiton replied:
Ah yes, I didn't see there was several commits, indeed https://github.com/twbs/bootstrap/pull/39483/commits/5d2fc6736d686e62268948e1cda51399f12ed71c looks even better this way. I couldn't make it the same for every padding from the form family neither.
Since we already have different variables for select and inputs in the rest of the code, maybe we could consider having separate variables in here as well.

louismaximepiton reviewed (changes requested):
I've been trying some other ideas that came to my mind but I couldn't find any better way to do it. I'm fine adding the Sass variables, I'm a bit more worried about the CSS variables that increase the size of the bundle.

Maybe we could calculate the paddings using the existing variables but I'm not sure it would be very handy as well.

Just a quick correction on the line-height to keep the correction very specific.

louismaximepiton reviewed (approved):
Fine with the actual on my side since it's better than it was. Still not a big fan of the line-height tweak since it introduces a default scroll for textareas but can't think of another way for now.

mdo commented:
Moving to v5.4.0 because new vars.
